Event Partitioning

EDsCODE opened the pull request:
## Changes
- migrates posthog_event table to a partitioned table by day
- celery worker keeps table updated by checking daily to make sure the partitions are a week ahead of current day

## TODO:
- [x] Modify update function to repair partitioning by moving data from default into proper partitions
- [x] FIX and add tests
- [x] handle postgres version differences

## Checklist
- [x] All querysets/queries filter by Team (if applicable)
- [x] Backend tests (if applicable)

EDsCODE committed:
initial working

EDsCODE committed:
update day to midnight

EDsCODE committed:
update CI postgres version

EDsCODE committed:
reverse migration file

timgl commented:
[Per discussion on Slack](https://posthog.slack.com/archives/C0113360FFV/p1588868567014000) we will silently ignore this migration if the database is Postgres 11. In that case, it will continue being un-partitioned.

EDsCODE committed:
handle default

EDsCODE committed:
add version check

EDsCODE committed:
add existence test and fixed bug

EDsCODE committed:
extend initial range partitioned

EDsCODE commented:
Ready for review. I did my own QA and the trend related queries along with paths query perform in <3 seconds on latest image of hosted instance. The migration took ~1-2minutes. Funnel query performance is around current performance (the ones on the dashboard take upwards of 5+ seconds). #695 + more optimizing on funnels should help this

EDsCODE committed:
reverse order of checks

timgl reviewed (commented):
Tested
- stopping query mid-migration
- migrating backwards and forwards
- This query performs the same (it's 30 days so needs to seq scan 30 tables): 
![image](https://user-images.githubusercontent.com/1727427/81351650-c1544c00-90bc-11ea-93cf-51efb50a6d50.png)
- Queries for 7 days with a filter perform about 50% better

Thoughts
- The migration takes around 8 minutes for 4 million events on my laptop. I left a comment that we could maybe look at the index building
- Someone like Maxim would have 5 million events in each table. When seq scanning for particular filters, I think this will be too slow. We discussed partitioning by event, which we might need to do. We already store all the event names against Team. If there are any new event names that day we could just store it in a 'other' bucket. What do you think?

EDsCODE commented:
Update: Will try the index deferring in the morning. I gave the event sub partitioning a go tonight in two different ways: (1) partioning on the first character as suggested in slack or (2) the entire event name. 

(1) doesnt appear likely to work because while the partioning itself works, a unique index can’t be created on the string shortening expression 

(2) should work but I was running into trouble with postgres automatically truncating super long strings (the hosted instance has some event names that are entire query strings or serialized dicts). Will continue this option

timgl commented:
Another thing I noticed was that going to an individual users' page and seeing all the events takes 50 seconds in a particular case where the user has about 15 distinct ids. 
![image](https://user-images.githubusercontent.com/1727427/81397720-c7c7df80-911f-11ea-8b8f-8b224f5dfc69.png)

Rewriting that query to get the distinct ids dynamically takes it down to 500ms
![image](https://user-images.githubusercontent.com/1727427/81397776-e62ddb00-911f-11ea-86e7-54e706555033.png)

Clicking on people in events was also very slow, which is probably worse because of this PR. I raised #736 to fix that

EDsCODE commented on posthog/celery.py:
```diff
@@ -26,12 +28,19 @@
+@app.task
+def update_event_partitions():
+    with connection.cursor() as cursor:
+        cursor.execute("DO $$ BEGIN IF (SELECT exists(select * from pg_proc where proname = 'update_partitions')) THEN PERFORM update_partitions(); END IF; END $$")
```
will run an empty migration and do nothing

EDsCODE replied:
i made the task check for function existence so it should do nothing if the function doesn't exist (migration didn't run because pg < 12)

timgl replied:
That was pretty obvious, thanks!

EDsCODE commented on posthog/migrations/sql/0050_event_partitions.sql:
```diff
@@ -0,0 +1,124 @@
+    EXECUTE ('CREATE INDEX posthog_event_team_id_a8b4c6dc ON public.posthog_event USING btree (team_id)');
+    EXECUTE ('CREATE INDEX posthog_event_idx_distinct_id ON public.posthog_event USING btree (distinct_id)');
+    EXECUTE ('CREATE INDEX posthog_eve_element_48becd_idx ON public.posthog_event USING btree (elements_hash)');
+    EXECUTE ('CREATE INDEX posthog_eve_timesta_1f6a8c_idx ON public.posthog_event USING btree ("timestamp", team_id, event)');
```
can't defer an index unless you mean concurrently but we decided that might be too drawn out

EDsCODE committed:
change granularity to week

EDsCODE commented:
I tested intervals of a week rather than a day. The granularity seemed to be making the append operation on the query suffer. All following results are testing on a table with 8million events

**RE: event + date filtering:** I managed to make the migration for this work however using the first letter of event partition. A lot more erroring because the table needs many more locks than the default setting allows. (90 days x 20-30 characters = too many tables). I can continue testing different combinations. Event partitioning over week might work 

**RE my results when filtering 90 days with browser property filter:** 40seconds with no partitioning, 3-20 seconds with week partioning. The 20seconds happened on initial query.

**RE other two issues:** I'm not quite sure how to reproduce as my queries on events, filtering events, clicking people and filtering were normal (1-3 seconds).

timgl committed:
Speed up loading of events with person_id

timgl commented:
This is great, comparable queries are down a few seconds more each.

I've pushed a fix for the slow loading of events for people. I think I'm happy to merge this now and deploy it on monday.

EDsCODE committed:
adjusted time to trigger and stay 2 weeks ahead instead of 1

timgl commented:
I do think it's worth doing event partitioning too. If you look at someone like Maxim, he'd still have 40 million events in one partition.

EDsCODE committed:
use commands instead of migration

EDsCODE committed:
update reverse script

EDsCODE commented:
Update: 
- removed migration
- moved sql into management command ex: `python manage.py partition --element '$pageview' --element '$autocapture'` or `python manage.py partition --reverse`
-added optional partitioning on certain events

TODO:
- [ ] test performance
- [ ] add tests 
- [ ] update docs

EDsCODE committed:
remove migration test

EDsCODE committed:
rebased master

EDsCODE committed:
adjust update function to handle all partitioned

EDsCODE committed:
handle errors

EDsCODE committed:
merge master

EDsCODE commented:
@timgl  Ready to test again. One extra functionality that probably should be added (maybe in a followup pr) is management command to add more partitions after the initial setup.

EDsCODE committed:
change the check that is performed

timgl commented:
@EDsCODE Great work on this and sorry it took so long.

Would you mind adding instructions to docs at posthog.com?

EDsCODE commented:
Will do
